src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -134,36 +146,43 @@ object DemuxFastqs {
     }
   }
 
-  /** Creates a demultiplexing iterator that performs demultiplexing in parallel.
+  /** Creates a demultiplexing iterator.
     *
     * @param sources the FASTQ sources, one per read.
     * @param demultiplexer the demultiplexer to use.  The demultiplexer's [[com.fulcrumgenomics.fastq.FastqDemultiplexer.demultiplex()]]
     *                      method expects the same number of reads as sources.
     */
-  def demultiplexingIterator(sources: Seq[FastqSource],
+  def demultiplexingIterator(sources: Seq[Iterator[FastqRecord]],
                              demultiplexer: FastqDemultiplexer,
-                             threads: Int,
-                             batchSize: Int = DemuxBatchRecordsSize): Iterator[DemuxResult] = {
-
+                             asyncOptions: DemuxFastqsAsyncOptions): Iterator[DemuxResult] = {
     require(demultiplexer.expectedNumberOfReads == sources.length,
       s"The demultiplexer expects ${demultiplexer.expectedNumberOfReads} reads but ${sources.length} FASTQ sources given.")
 
-    val zippedIterator = FastqSource.zipped(sources)
-    if (threads > 1) {
+    val demuxSource = asyncOptions.asyncReadingType match {
+      case DemuxFastqsAsyncReadingType.NoAsync         => FastqSource.zipped(sources)
+      case DemuxFastqsAsyncReadingType.SingleThread    => AsyncIterator(FastqSource.zipped(sources), bufferSize=Some(1024))
+      case DemuxFastqsAsyncReadingType.ThreadPerReader => FastqSource.zipped(sources.map { source => AsyncIterator(source, bufferSize=Some(1024))})
+    }
+
+    if (asyncOptions.asyncDemux) {
+      AsyncIterator[DemuxResult](demuxSource.map { readRecords => demultiplexer.demultiplex(readRecords: _*) }, bufferSize=Some(asyncOptions.demuxBatchSize))
+      // NB: We could enable multiple threads for demultiplexing with this in the future.
       // Developer Note: Iterator does not support parallel operations, so we need to group together records into a
       // [[List]] or [[Seq]].  A fixed number of records are grouped to reduce memory overhead.
+      /*
       import com.fulcrumgenomics.FgBioDef.ParSupport
-      val pool = new ForkJoinPool(threads)
-      zippedIterator
-        .grouped(batchSize)
+      val pool = new ForkJoinPool(demuxThreads)
```

Comment:
Does using > 1 extra thread here help throughput significantly?  If not you could just wrap this in an AsyncIterator, but I'm guessing more threads must be worth the complexity of batching?